### DISCUSSION TITLE
Improving support for JSON format validation of strings with search patterns

### DISCUSSION THREAD

**Discussion by Markus-Elfring** [json-schema-org/community#218] [Q&A] [answered]
The JSON schema specification is providing the functionality to [perform validation of strings with search patterns](https://json-schema.org/understanding-json-schema/reference/string.html#regular-expressions "Description for the usage of the pattern keyword").
Thus I came along to apply this feature for another simple data type also together with the parser generator “jsonschema2pojo 1.1.2”.
Unfortunately, I stumbled on the development situation that [this tool does not support the desired checking of the data type “string” by the means of regular expressions](https://github.com/joelittlejohn/jsonschema2pojo/issues/1419 "missing Java code generation") (at the moment).

:crystal_ball: Now I am curious how the chances will evolve for better support of corresponding techniques (according to related software tools).

**Comment by gregsdennis** [comment, marked as answer]
@Markus-Elfring are you looking to improve that tool or the specification?  Do you have a specific improvement in mind?

> better support of corresponding techniques

I'm not sure what you're asking for.  Can you give some examples?

**Reply by Markus-Elfring** [reply]
>are you looking to improve that tool or the specification?

I am mostly looking for enhancement possibilities according to affected tools.
:crystal_ball: Will any additional development efforts influence the software evolution in desirable directions?


>I'm not sure what you're asking for.

How good did you understand the information sources which I linked for my clarification approach?


>Can you give some examples?

I am curious if the data format handling will be clarified further for a JSON example like the following.
```JSON
{
  "$schema": "http://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/GTIN-13",
  "type": "string",
  "pattern": "^[0-9]{13}$"
}
```

**Reply by gregsdennis** [reply]
The specification is pretty clear on the _validation_ requirements around `pattern`, but that doesn't apply to code generation, which is what the tool you linked to does.  JSON Schema doesn't have _any_ requirements around code generation (although there is [an effort](https://github.com/json-schema-org/vocab-idl) to define something similar).

I suspect that your use of the word "validation" in the issue you linked made the author think that validation what you were looking for, but it sounds to me like you wanted some Java model code as output.

Is that right?  If so, I'd be happy to help you in that issue.

**Reply by Markus-Elfring** [reply]
>The specification is pretty clear on the _validation_ requirements around `pattern`, but that doesn't apply to code generation, …

Thanks for your distinction attempt.

I hope that corresponding parser generators can help more with desired data processing for various input files.


>…, but it sounds to me like you wanted some Java model code as output.

* Should any generated source code take care of the proper data format checks?
* [Do you know any programming tools](https://json-schema.org/implementations.html "Evolving tool support") which can process relevant data (based on JSON schema specifications) without code generation?

**Reply by Relequestual** [reply]
JSON Schema the organization has no control over implementations.

> Should any generated source code take care of the proper data format checks?

JSON Schema makes no statements about any generated source code. Each library can do anything or nothing, they are all different. It is out of scope for the current specification documents.

> [Do you know any programming tools](https://json-schema.org/implementations.html) which can process relevant data (based on JSON schema specifications) without code generation?

Could you explain what you mean by this please? If you mean, validate the data, then all the Java implementations we know about can be found on the page you linked, under validators, Java: https://json-schema.org/implementations.html#validator-java

**Reply by Markus-Elfring** [reply]
>>[Do you know any programming tools](https://json-schema.org/implementations.html "Evolving tool support") which can process relevant data (based on JSON schema specifications) without code generation?

>Could you explain what you mean by this please?

* One of the data processing approaches reuses parser generators (which will usually generate some code), doesn't it?
* Another approach would import JSON schema specifications into database variants for subsequent data analyses, wouldn't it?


>If you mean, validate the data, …

I would like to achieve that not only validation will be performed for selected input data but also further processing according to extracted information.

:thought_balloon: Is it unclear how good the linked validators collaborate with parsing functionality?

**Reply by Relequestual** [reply]
> ... further processing according to extracted information.

I appreciate there is a language barrier here, but I'm afraid I don't understand what this means.
Maybe you could give an example of what you mean?

**Reply by Markus-Elfring** [reply]
>I appreciate there is a language barrier here,

Maybe?


>but I'm afraid I don't understand what this means.

[Data validation corresponds to a known meaning](https://en.wikipedia.org/wiki/Data_validation "Description for data validation").
:thought_balloon: Its functionality depends also on parsing for required data extraction, doesn't it?
Desirable quality of service properties are evolving accordingly.

**Reply by Relequestual** [reply]
I'm pretty familiar with "data validation", which isn't what you said.

However...

> 💭 Its functionality depends also on parsing for required data extraction, doesn't it?
> Desirable quality of service properties are evolving accordingly.

I'm afraid I don't understand what those words in that order means.
Maybe you could ask a colleague to help you rephrase this?

I'm sorry, but myself, and others, are finding it very hard to understand your sentences. Are you using some sort of automated translation tool?

**Reply by handrews** [reply]
@Markus-Elfring The JSON Schema organization produces the specification, test suite data, and our website (now including the Understanding JSON Schema site).

_We do not produce validation/code generation/document generation or any other implementations of the specifcation_.

This is not the place to ask about such things.  Looking at the jsonschema2pojo issue you linked, the problem is not with the specification.  They understand the specification, they have simply excluded validation from their project.

We cannot do anything about that here.  Their advice for you to find a validator that does what you want is already the correct answer.

**Reply by Markus-Elfring** [reply]
>>:thought_balloon: Its functionality depends also on parsing for required data extraction, doesn't it?
>>Desirable quality of service properties are evolving accordingly.

>I'm afraid I don't understand what those words in that order means.

I find such communication difficulties interesting somehow.


>I'm sorry, but myself, and others, are finding it very hard to understand your sentences.

I hope that involved understanding challenges can still be resolved.


>Are you using some sort of automated translation tool?

Not for the discussion according to this issue.

**Reply by Markus-Elfring** [reply]
>This is not the place to ask about such things.

I am trying to clarify information a bit more also according to evolving software tool support.


>They understand the specification,

I hope so.


>they have simply excluded validation from their project.

:thought_balloon: I imagine that such an interpretation of the referenced siutation should trigger further development considerations.


>We cannot do anything about that here.

:thought_balloon: I guess that growing interests by some software users for involved issues can influence implementation details into more desirable directions.


>Their advice for you to find a validator that does what you want is already the correct answer.

I propose to reconsider data processing requirements accordingly.

Tool alternatives have got also their remaining open issues (as usual).

**Reply by Markus-Elfring** [reply]
>…, the problem is not with the specification.

I stumbled on another development concern.
I found a file like [“ecmascript-regex.json” in the JSON Schema Test Suite](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/e42e8417b6094c613a7bd6a2f1fbbbf48ade66ff/tests/draft2020-12/optional/ecmascript-regex.json#L3 "Checking the usage of patterns").
:crystal_ball: Will a need evolve to reconsider [the selection of the subdirectory “optional”](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/2a316683645441e5966b71f72c08f3c6e9843dca/README.md#subdirectories-within-each-draft "Description for optional test items") for its storage?

**Reply by handrews** [reply]
> > I'm afraid I don't understand what those words in that order means.
> 
> I find such communication difficulties interesting somehow.

> > I'm sorry, but myself, and others, are finding it very hard to understand your sentences.
> 
> I hope that involved understanding challenges can still be resolved.

That will require you to be more clear.  Referring to communication challenges as "interesting" rather than finding a more direct way (perhaps, assuming this is a language difficulty, with the assistance of another person who speaks both your native language and English?) is not going to lead to a resolution.

I don't see a way forward for this discussion without a much more clear set of questions.  If you cannot provide that, I will likely lock this discussion tomorrow.  We would like to help, but we just don't understand what you're asking.

**Reply by Markus-Elfring** [reply]
* Are parser generators available (for JSON schemas) which support the evaluation of search patterns in standard-compliant ways?
* Does the corresponding handling of regular expressions belong to mandatory software functionality?

**Reply by handrews** [reply]
> Are parser generators available (for JSON schemas) which support the evaluation of search patterns in standard-compliant ways?

We have a [list of known implementations](https://json-schema.org/implementations.html) on our web page.

> Does the corresponding handling of regular expressions belong to mandatory software functionality?

The core specification has a section on [what is required regarding regular expressions](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-regular-expressions).
